# Min-Max algorithm implemetation for Gomoku player

__Define a random player to play against min-max__

In [0]:
from util import print_board
from GomokuBoard import GomokuBoard, GameResult
from GamePlayerInterface import GamePlayerInterface

class RandomPlayer(GamePlayerInterface):

    def __init__(self):
        self.side = None
        super().__init__()

    def move(self, GomokuBoard : GomokuBoard) -> (GameResult, bool):

#         print('move to', GomokuBoard .random_empty_spot())
        location = input("Your move: ")
        _, res, finished = GomokuBoard .move(int(location), self.side)
        
        print_board(GomokuBoard )

        return res, finished

    def final_result(self, result: GameResult):
        pass

    def new_game(self, side: int):
        self.side = side

In [0]:
from IPython.display import HTML, display
from GomokuBoard  import GomokuBoard, GameResult, CROSS, NAUGHT, EMPTY
from util import print_board, play_game

__We use an 8*8 board for Gomoku__

In [0]:
board = GomokuBoard()
player1 = RandomPlayer()
player2 = RandomPlayer()

In [0]:
result = play_game(board, player1, player2)

Your move: 11
self.state [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
position 11


In [0]:
print_board(board)

if result == GameResult.CROSS_WIN:
    print("Cross won")
elif result == GameResult.NAUGHT_WIN:
    print("Naught won")
else:
    print("Draw")

__Min-Max vs Random player(Min-max player go first)__

In [0]:
from GamePlayerInterface import GamePlayerInterface

def battle(player1: GamePlayerInterface, player2: GamePlayerInterface, num_games: int = 100000):
    board = GomokuBoard()
    draw_count = 0
    cross_count = 0
    naught_count = 0
    for _ in range(num_games):
        result = play_game(board, player1, player2)
        if result == GameResult.CROSS_WIN:
            cross_count += 1
        elif result == GameResult.NAUGHT_WIN:
            naught_count += 1
        else:
            draw_count += 1

    print("After {} game we have draws: {}, GamePlayerInterface 1 wins: {}, and GamePlayerInterface 2 wins: {}.".format(num_games, draw_count,
                                                                                         cross_count, naught_count))

    print("Which gives percentages of draws: {:.2%}, GamePlayerInterface 1 wins: {:.2%}, and GamePlayerInterface 2 wins:  {:.2%}".format(
        draw_count / num_games, cross_count / num_games, naught_count / num_games))

In [0]:
from MinMaxAlgorithm import MinMaxAlgorithm

battle(MinMaxAlgorithm(), RandomPlayer())

After 100000 game we have draws: 489, GamePlayerInterface 1 wins: 99511, and GamePlayerInterface 2 wins: 0.
Which gives percentages of draws: 0.49%, GamePlayerInterface 1 wins: 99.51%, and GamePlayerInterface 2 wins:  0.00%


__Min-Max vs Random player(Random player go first)__

In [0]:
battle(RandomPlayer(), MinMaxAlgorithm())

After 100000 game we have draws: 19323, GamePlayerInterface 1 wins: 0, and GamePlayerInterface 2 wins: 80677.
Which gives percentages of draws: 19.32%, GamePlayerInterface 1 wins: 0.00%, and GamePlayerInterface 2 wins:  80.68%


__2 MinMax players against each other__

In [0]:
battle(MinMaxAlgorithm(), MinMaxAlgorithm())

After 100000 game we have draws: 100000, GamePlayerInterface 1 wins: 0, and GamePlayerInterface 2 wins: 0.
Which gives percentages of draws: 100.00%, GamePlayerInterface 1 wins: 0.00%, and GamePlayerInterface 2 wins:  0.00%


we get 100% draws.

# Summary

This gives us an additional baseline:

| GamePlayerInterface | P1 Win | P2 Win | Draw |
| --- | ---| --- | --- |
| Min Max - Random | 99.5%     | 0%     | 0.5% |
| Random - Min Max | 0%    | 80% | 20% |
| Min Max - Min Max | 0%    | 0% | 100% |

This means, in order to be considered as playing better than pure random, a player should achieve significantly more than 20% draws against a Min-Max player when going first and significantly more than 1% draws when going second.
